Search all view paths for requested template

jamis opened the pull request:
This is a simple patch, iterating over all supplied bases until a requested template is found. It allows Cell::ViewModel.view_paths to be appended to, so that (for instance) engines can bundle cells.

(Related: the docs seem to suggest that cells can be bundled in engines, but maybe that was pre-4.0? At any rate, I couldn't get it to work in 4.0 without the changes in this patch.)

I'm embarrassed that there is no test to accompany this patch, but I'll be honest---I spent a good hour trying to get the tests to run, and failed miserably. If you can point me at instructions for getting the dev environment configured, I'd be happy to work some tests into this patch, but as it is...a guy's gotta get work done to pay the bills, you know? :)

apotonick commented:
As you can tell, there's a `FIXME` that is supposed to remind me to go through all base paths.

Thanks for your patch, but we have to restructure that a bit so we make use of caching.

What's the deal with the tests? Not running for you? And, yeah, I know what you mean with "paying the bills" *cough* :stuck_out_tongue_winking_eye:

jamis commented:
Yeah, I figured there was a reason the FIXME was there instead of something like my (nearly trivial) patch. :) As for the tests, after running `bundle`, and then `bundle exec rake`, I get `cannot load such file -- rails` -- which seems very odd. So I notice that Gemfile.lock is missing rails, so I add `gem "rails"` to Gemfile and try again, and...same error.

HOWEVER. Trying it all again just now, I notice that Gemfile.lock now has a dependency on rails 0.9.5 (wha?), so I change Gemfile to say `gem "rails", "4.2.0"`, and rebundle, and now...

Hey. It works. :P So, maybe I'm doing it wrong, but it seems like your Gemfile might need a dependency on rails itself, and not just actionpack? (Maybe this is the fault of rails 4.2.0?)

apotonick commented:
Yeah, sorry, the `Gemfile` is wrong, I always run tests with the newest rails version from `gemfiles`. Sorry haha. Should all be good and running in a few weeks, can't wait to release Cells 4!!

jamis commented:
Cool, no worries. :) I'll live off my fork for now, so I can have the engine loading, and I'll sit tight and wait patiently for the other goodies. :) If I have a chance, I'll have a look tomorrow at considering caching in this patch.

jamis commented:
Alright, I moved the caching from Temlates#find_template to Templates#[], to accommodate the wider net cast by the bases check. I also added a test, and confirmed all tests are still passing. (yay!) Does this fit with what you meant by "make use of caching?"

bmcdaniel11 commented:
Is there any way we can get this merged into the latest release? It's a roadblock for engines.

Thanks!

apotonick commented:
Yes, but we also need a test setup for testing engine cells and this is some work. I would appreciate if someone could help me with the engine test.

apotonick commented:
I have several questions on how you use cells in/with engines.

1. If your engine bundles cells that should simply be usable across the app, you could simply reset the `view_paths` and it would still work with the current (insufficient) implementation

    ```ruby
    MyEngine::Cell.view_paths = "/var/engines/app/cells/"
    ```
2. The problem arises when you _inherit_ from engine cells  in another directory (do you guys do that?).

    ```ruby
    class SongCell < MyEngine::Cell
    ```

    Now that application cell needs two bases, namely `Rails.root/app` (application) and `/var/engines/...`.

I'm wondering if we should merge prefixes and view_paths. There's simply no such case where you'd want Cells to look for `/var/engines/cells/song/show.haml` (application cell in engine, it's never gonna be there). Likewise, you don't want it to search for `Rails.root/app/cells/myengine/show.haml` because overriding engine views in the application namespace without actually inheriting from the engine is super dirty.

Thoughts?

jamis commented:
The issue I had was with the second point you made, where the engine contained a base cell class, and the application inherited from it. We were doing this because we had a gem that encapsulated logic to be shared between several related apps (think "service-oriented architecture"). I've been out of that project for several months now, so it's not very fresh in my mind, but I do recall that we also had shared partials -- bits that subclassed cells could take advantage of in order to build a custom view. I'm not sure if that falls under the "no such case where you'd want Cells to look for /var/engines/cells/song/show.haml" bit.

apotonick commented:
@jamis Partials, or cell views?

jamis commented:
Partials only. Concrete cell views themselves were always defined where cells themselves were defined (app or engine).

apotonick commented:
You were rendering partials from cells? With Cells 4?

I totally get subclassing engine cells and overriding views - that's why we introduced view inheritance in the first place.

apotonick commented:
I will most probably merge `::prefixes` and `::view_paths` do avoid confusion and simplify the entire process of view finding.

1. Every cell can have one view path.

    ```ruby
    MyEngine::Cell.view_path = "/var/my_engine/cells"

    SongCell.view_path = "app/cells"
    ```
2. `::prefixes` will contain the base path already, so `ViewModel#find_template` doesn't have to figure it out.

    ```ruby
    MyEngine::Cell.prefixes #=> ["/var/my_engine/cells/my_engine"]
    SongCell.prefixes  #=> ["app/cells/song"]
    ```
3. When inheriting, we simply inherit `::prefixes` from the parent cell, the way we do it now.

    ```ruby
    HitCell < MyEngine::Cell
    HitCell.prefixes #=> ["app/cells/hit", "/var/my_engine/cells/my_engine"]
    ```

4. If you're not happy with that, you can push more onto `::prefixes`

    ```ruby
    class MegaHitCell < Cell::ViewModel
      def self.prefixes
        super + SongCell.prefixes
      end
    end

    MegaHitCell.prefixes #=> ["app/cells/hit", "app/cells/song"]
    ```

This greatly reduces the lookup algorithm complexity and makes it simply to understand where views come from, no need to explain "it first looks through base paths, then the prefixes, then the other base path, ..." and so on. Cool?

jamis commented:
That sounds great to me!

(FWIW, no, we weren't using cells 4. The previous architecture was using cells 3, and I was trying to upgrade us, but as this thread has noted, there were...obstacles. In the end, the project got mothballed anyway, so all I've *really* got is "we wanted it to work like X". Apologies for not being clearer.)

apotonick commented:
@jamis Thanks for your invalueable help, man! I decided to implement the above solution and did not merge your PR. Anyway, feel free to PR again, I will merge it! :heart: :beers:

apotonick commented:
I started documenting here: http://trailblazerb.org/gems/cells/render.html

bmcdaniel11 commented:
Thanks, Nick, your solution looks like it will work fine. Just for documentation sake, I too am using the engine cell in the same way jamis is, minus the rendering partials. I have a base cell in the engine I inherit from for several of my cells. Appreciate the work you are doing, keep up the good work!

jamis commented:
:+1:

apotonick commented:
@bmcdaniel11 Some feedback here would be amazing: https://github.com/apotonick/cells/issues/280
